openverse_catalog/dags/common/sensors.py

```diff
@@ -0,0 +1,93 @@
+import os
+from datetime import timedelta
+from typing import Iterable
+
+from airflow.exceptions import AirflowException
+from airflow.models import DagModel, DagRun
+from airflow.sensors.base import BaseSensorOperator
+from airflow.triggers.temporal import TimeDeltaTrigger
+from airflow.utils.session import provide_session
+from airflow.utils.state import State
+from sqlalchemy import func
+
+
+class ExternalDAGsSensorAsync(BaseSensorOperator):
+    """
+    Waits for a list of different DAGs to not be running, freeing up
+    a worker slot while it's waiting.
+
+    external_dag_ids: A list of dag_ids that you want to wait for
+    :param check_existence: Set to `True` to check if the external task exists (when
+        external_task_id is not None) or check if the DAG to wait for exists (when
+        external_task_id is None), and immediately cease waiting if the external task
+        or DAG does not exist (default value: False).
+    """
+
+    # TODO: Not sure if this is necessary
+    template_fields = ["external_dag_ids"]
+
+    # TODO: I don't understand what operator extra links are, and whether it's
+    # necessary to implement this.
+    # @property
+    # def operator_extra_links(self):
+    #     """Return operator extra links"""
+    #     return [ExternalDAGsSensorLink()]
+
+    def __init__(
+        self,
+        *,
+        external_dag_ids: Iterable[str],
+        check_existence: bool = False,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.external_dag_ids = external_dag_ids
+        self.check_existence = check_existence
+        self._has_checked_existence = False
+
+    @provide_session
+    def execute(self, context, event=None, session=None):
```

Comment:
I think `BaseSensorOperator::poke` is the function we want to override!

https://github.com/apache/airflow/blob/78490f86bfab195cc98a7b700c26c24a1aa2eb17/airflow/sensors/base.py#L149-L154

Reply:
I think this is how I had to do it with Deferrable sensors -- I can't believe I missed you can free the worker slot with 'normal' sensors! Thanks, changed it back to `poke` and added `reschedule` mode. So much nicer not to have to deal with Triggerers, amazing 🎉 😄

Reply:
I think the deferrable stuff is pretty new and (frankly) _very_ cool, but it solves a problem we (fortunately) don't yet have haha